library/asn1write.c

```diff
@@ -287,26 +287,83 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
     return( (int) len );
 }
 
+int mbedtls_asn1_write_named_bitstring( unsigned char **p,
+                                        unsigned char *start,
+                                        const unsigned char *buf,
+                                        size_t bits )
+{
+    size_t unused_bits, byte_len, input_byte_len;
+
+    /*
+     * Named bitstrings require that trailing 0s are excluded in the encoding
+     * of the bitstring. Trailing 0s are considered part of the 'unused' bits
+     * when encoding this value in the first content octet
+     */
+    byte_len = ( bits / 8 ) + ( ( bits % 8 ) ? 1 : 0 );
+    input_byte_len = byte_len;
+
+    /* Scan the input data from back to front and remove trailing 0s */
+    for( ; byte_len > 0; byte_len-- )
+    {
+        if( buf[byte_len - 1] == 0 )
+        {
+            continue;
+        }
+
+        unused_bits = ( input_byte_len == byte_len ) ?
+            ( 8 -  ( bits % 8 ) ) % 8 : 0;
+
+        for( ; unused_bits < 8; unused_bits++ )
+        {
+            if( ( ( buf[byte_len - 1] >> unused_bits ) & 0x1 ) != 0 )
+            {
+                break;
+            }
+        }
+
+        if( unused_bits == 8 )
+        {
+            /* Ensure that the leading byte is 0 if the bitstring is empty */
+            unused_bits = 0;
```

Comment:
I might misunderstand the code, but this assignment is redundant, isn't it? `unused_bits` is used for the computation of `bits` only if `byte_len > 0`, which can only happen if the loop was exited through the `break;` instruction at the end of its body, in which case `unused_bits` has been assigned in that iteration.

I'd find the code easier to read if it followed the following structure:
(1) Read `buf[byte_len - 1]` into a temporary `unsigned char`.
(2) Mask away the unused bits if `byte_len == input_byte_len`.
(3) Continue if it's `0`. 
(4) Otherwise, count unused bits bits and abort.